File: config/src/main/scala/net/kemitix/thorp/config/ConfigQuery.scala
package net.kemitix.thorp.config

import java.nio.file.Paths
import scala.jdk.CollectionConverters._

import net.kemitix.thorp.domain.Sources

trait ConfigQuery {

  def showVersion(configOptions: ConfigOptions): Boolean =
    ConfigOptions.contains(ConfigOption.Version)(configOptions)

  def batchMode(configOptions: ConfigOptions): Boolean =
    ConfigOptions.contains(ConfigOption.BatchMode)(configOptions)

  def ignoreUserOptions(configOptions: ConfigOptions): Boolean =
    ConfigOptions.contains(ConfigOption.IgnoreUserOptions)(configOptions)

  def ignoreGlobalOptions(configOptions: ConfigOptions): Boolean =
    ConfigOptions.contains(ConfigOption.IgnoreGlobalOptions)(configOptions)

  def sources(configOptions: ConfigOptions): Sources = {
    val explicitPaths = configOptions.options.asScala.flatMap {
      case ConfigOption.Source(sourcePath) => List(sourcePath)
      case _                               => List.empty
    }.toList
    val paths = explicitPaths match {
      case List() => List(Paths.get(System.getenv("PWD")))
      case _      => explicitPaths
    }
    Sources.create(paths.asJava)
  }

}

object ConfigQuery extends ConfigQuery

File: config/src/main/java/net/kemitix/thorp/config/ConfigQuery.java
package net.kemitix.thorp.config;

import net.kemitix.thorp.domain.Sources;

import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

public interface ConfigQuery {

    static boolean showVersion(ConfigOptions configOptions) {
        return configOptions.options().stream()
                .anyMatch(configOption ->
                        configOption instanceof ConfigOption.Version);
    }

    static boolean batchMode(ConfigOptions configOptions) {
        return configOptions.options().stream()
                .anyMatch(configOption ->
                        configOption instanceof ConfigOption.BatchMode);
    }

    static boolean ignoreUserOptions(ConfigOptions configOptions) {
        return configOptions.options().stream()
                .anyMatch(configOption ->
                        configOption instanceof ConfigOption.IgnoreUserOptions);
    }

    static boolean ignoreGlobalOptions(ConfigOptions configOptions) {
        return configOptions.options().stream()
                .anyMatch(configOption ->
                        configOption instanceof ConfigOption.IgnoreGlobalOptions);
    }

    static Sources sources(ConfigOptions configOptions) {
        List<Path> explicitPaths = configOptions.options().stream()
                .filter(configOption ->
                        configOption instanceof ConfigOption.Source)
                .map(ConfigOption.Source.class::cast)
                .map(ConfigOption.Source::path)
                .collect(Collectors.toList());
        if (explicitPaths.isEmpty()) {
            return Sources.create(Collections.singletonList(Paths.get(System.getenv("PWD"))));
        }
        return Sources.create(explicitPaths);
    }
}
